File: plog/src/main/java/com/blackbox/plog/PLog.java
package com.blackbox.plog;

/**
 * Created by Umair Adil on 12/04/2017.
 */

import android.util.Log;

import com.blackbox.plog.utils.DateControl;
import com.blackbox.plog.utils.DateTimeUtils;
import com.blackbox.plog.utils.Utils;

import org.apache.log4j.Logger;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.io.IOException;
import java.util.concurrent.Callable;
import java.util.zip.ZipEntry;
import java.util.zip.ZipOutputStream;

import de.mindpipe.android.logging.log4j.LogConfigurator;
import io.reactivex.Observable;

public class PLog {

    private static String TAG = PLog.class.getSimpleName();
    private final static LogConfigurator _logConfigurator = new LogConfigurator();

    private static String zipName = "";
    protected static PLogger pLogger = new PLogger();

    //Log Filters
    public static final int LOG_TODAY_ONLY = 0;
    public static final int LOG_TODAY = 1;
    public static final int LOG_LAST_HOUR = 2;
    public static final int LOG_WEEK = 3;
    public static final int LOG_LAST_TWO_DAYS = 4;

    //Log Types
    public static final String TYPE_INFO = "Info";
    public static final String TYPE_ERROR = "Error";
    public static final String TYPE_WARNING = "Warning";

    static void setPLogger(PLogger pLog) {
        pLogger = pLog;
    }

    /**
     * Configure Log4j
     *
     * @param fileName      Name of the log file
     * @param filePattern   Output format of the log line
     * @param maxBackupSize Maximum number of backed up log files
     * @param maxFileSize   Maximum size of log file until rolling
     */
    private static void Configure(String fileName, String filePattern,
                                  int maxBackupSize, long maxFileSize, boolean logEvent) {

        try {
            _logConfigurator.setFileName(fileName);
            _logConfigurator.setFilePattern(filePattern);
            _logConfigurator.setMaxBackupSize(maxBackupSize);
            _logConfigurator.setMaxFileSize(maxFileSize);
            _logConfigurator.setInternalDebugging(false);
            _logConfigurator.setUseLogCatAppender(logEvent);
            _logConfigurator.configure();

        } catch (Exception e) {
            e.printStackTrace();
        }

    }

    public static void logThis(String className, String functionName, String text, String type) {
        try {
            //Make sure what is logged is unique
            if (!Utils.getInstance().isLoggedOnce(text)) {


                String folderPath = getLogPath() + DateControl.getInstance().getToday();
                Utils.getInstance().createDirIfNotExists(folderPath);

                String fileName_raw = DateControl.getInstance().getToday() + DateControl.getInstance().getHour();
                String path_raw = folderPath + File.separator + fileName_raw;
                boolean existsRaw = Utils.getInstance().checkFileExists(path_raw);

                if (!existsRaw) {
                    String filePattern = "%m%n";
                    int maxBackupSize = 2500;
                    long maxFileSize = 2048 * 2048;
                    Configure(path_raw, filePattern, maxBackupSize, maxFileSize, true);
                }

                LogData logData = new LogData(className, functionName, text, DateTimeUtils.getFullDateTimeString(System.currentTimeMillis()), type);

                String logFormatted = LogFormatter.getFormatType(logData, pLogger);

                Logger.getLogger(TAG).info(logFormatted);
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
    }


    protected static String getOutputPath() {
        return pLogger.getExportPath() + File.separator;
    }

    private static String getLogPath() {
        return pLogger.getSavePath() + File.separator;
    }

    public static Observable<String> getLogs(final int type) {

        return Observable.fromCallable(new Callable<String>() {
            @Override
            public String call() throws Exception {

                FileFilter.prepareOutputFile();

                String path = "";
                int files = 0;
                String timeStamp = "";
                String noOfFiles = "";


                switch (type) {
                    case LOG_TODAY_ONLY:
                        path = getLogPath() + DateControl.getInstance().getToday();
                        files = FileFilter.getFilesForToday(path);

                        if (pLogger.getAttachTimeStamp())
                            timeStamp = "_" + DateTimeUtils.getFullDateTimeStringCompressed(System.currentTimeMillis());

                        if (pLogger.getAttachNoOfFiles())
                            noOfFiles = "_[" + files + "]";

                        zipName = pLogger.getExportFileName() + timeStamp + noOfFiles + ".zip";
                        break;

                    case LOG_TODAY:

                        path = getLogPath() + DateControl.getInstance().getToday();
                        files = FileFilter.getFilesForToday(path);

                        if (pLogger.getAttachTimeStamp())
                            timeStamp = "_" + DateTimeUtils.getFullDateTimeStringCompressed(System.currentTimeMillis()) + "_Today";

                        if (pLogger.getAttachNoOfFiles())
                            noOfFiles = "_[" + files + "]";

                        zipName = pLogger.getExportFileName() + timeStamp + noOfFiles + ".zip";
                        break;

                    case LOG_LAST_HOUR:
                        path = getLogPath() + DateControl.getInstance().getToday();
                        FileFilter.getFilesForLastHour(path);

                        if (pLogger.getAttachTimeStamp())
                            timeStamp = "_" + DateTimeUtils.getFullDateTimeStringCompressed(System.currentTimeMillis()) + "_LastHour";

                        if (pLogger.getAttachNoOfFiles())
                            noOfFiles = "_[" + 1 + "]";

                        zipName = pLogger.getExportFileName() + timeStamp + noOfFiles + ".zip";
                        break;

                    case LOG_WEEK:
                        FileFilter.getFilesForLastWeek(getLogPath());

                        if (pLogger.getAttachTimeStamp())
                            timeStamp = "_" + DateTimeUtils.getFullDateTimeStringCompressed(System.currentTimeMillis()) + "_Week";

                        if (pLogger.getAttachNoOfFiles())
                            noOfFiles = "_[" + 1 + "]";

                        zipName = pLogger.getExportFileName() + timeStamp + noOfFiles + ".zip";
                        break;

                    case LOG_LAST_TWO_DAYS:
                        FileFilter.getFilesForLastTwoDays(getLogPath());

                        if (pLogger.getAttachTimeStamp())
                            timeStamp = "_" + DateTimeUtils.getFullDateTimeStringCompressed(System.currentTimeMillis()) + "_Last2Days";

                        if (pLogger.getAttachNoOfFiles())
                            noOfFiles = "_[" + 1 + "]";

                        zipName = pLogger.getExportFileName() + timeStamp + noOfFiles + ".zip";

                        break;

                }

                final String outputPath = getOutputPath();
                final File outputDirectory = new File(outputPath);
                final File[] filesToSend = outputDirectory.listFiles();

                if (filesToSend != null && filesToSend.length > 0) {
                    if (pLogger.isDebuggable())
                        logThis(TAG, "createZipFile", "Start Zipping Log Files.. " + filesToSend.length, TYPE_INFO);
                } else {
                    if (pLogger.isDebuggable())
                        logThis(TAG, "createZipFile", "No Files to zip!", TYPE_WARNING);
                    return null;
                }

                try {
                    FileOutputStream fos = null;
                    try {
                        fos = new FileOutputStream(outputPath + zipName);

                        ZipOutputStream zos = new ZipOutputStream(fos);

                        for (File f : filesToSend) {
                            if (!f.getName().contains(".zip")) {

                                if (pLogger.isDebuggable())
                                    logThis(TAG, "zipFile", "Adding file: " + f.getName(), TYPE_INFO);

                                byte[] buffer = new byte[1024];
                                FileInputStream fis = new FileInputStream(f);
                                zos.putNextEntry(new ZipEntry(f.getName() + ".txt"));
                                int length;
                                while ((length = fis.read(buffer)) > 0) {
                                    zos.write(buffer, 0, length);
                                }
                                zos.closeEntry();
                                fis.close();
                            }
                        }

                        zos.close();
                    } catch (FileNotFoundException e) {
                        e.printStackTrace();
                        if (pLogger.isDebuggable())
                            logThis(TAG, "getLogs", "Error: " + e.getLocalizedMessage(), TYPE_ERROR);
                        return null;
                    }
                } catch (IOException ioe) {
                    Log.e(TAG, ioe.getMessage());
                    if (pLogger.isDebuggable())
                        logThis(TAG, "getLogs", "Error: " + ioe.getLocalizedMessage(), TYPE_ERROR);
                    return null;
                }

                if (pLogger.isDebuggable())
                    logThis(TAG, "getLogs", "Output Zip: " + outputPath + zipName, TYPE_INFO);

                return outputPath + zipName;
            }
        });

    }


    /**
     * Clear logs boolean.
     * <p>
     * <p>Will return true if delete was successful</p>
     *
     * @return the boolean
     */
    public static boolean clearLogs() {
        return Utils.getInstance().deleteDir(new File(getLogPath()));
    }
}

File: plog/src/main/java/com/blackbox/plog/PLog.kt
package com.blackbox.plog

/**
 * Created by Umair on 12/04/2017.
 */

import android.os.Environment
import android.util.Log
import de.mindpipe.android.logging.log4j.LogConfigurator
import org.apache.log4j.Logger
import java.io.File

class PLog {

    companion object Factory {
        fun create(): PLog = PLog()
    }

    val dateControl = DateControl.create()
    val dateTime = DateTimeUtils.create()
    val compress = Compress.create()

    private val TAG = PLog::class.java.simpleName
    private val _logConfigurator = LogConfigurator()

    val TYPE_INFO = "Info"
    val TYPE_ERROR = "Error"

    val LOG_TODAY = 1
    val LOG_LAST_HOUR = 2
    val LOG_WEEK = 3

    /**
     * Configure Log4j
     * @param fileName      Name of the log file
     * *
     * @param filePattern   Output format of the log line
     * *
     * @param maxBackupSize Maximum number of backed up log files
     * *
     * @param maxFileSize   Maximum size of log file until rolling
     */
    private fun Configure(fileName: String, filePattern: String,
                          maxBackupSize: Int, maxFileSize: Long, logEvent: Boolean) {

        try {
            _logConfigurator.fileName = fileName
            _logConfigurator.filePattern = filePattern
            _logConfigurator.maxBackupSize = maxBackupSize
            _logConfigurator.maxFileSize = maxFileSize
            _logConfigurator.isInternalDebugging = false
            _logConfigurator.isUseLogCatAppender = logEvent
            _logConfigurator.configure()
        } catch (e: Exception) {
            e.printStackTrace()
        }

    }

    fun logThis(className: String, functionName: String, text: String, type: String) {
        try {

            val folderPath = logPath + dateControl.today

            Utils.instance.createDirIfNotExists(folderPath)


            val fileName_raw = dateControl.today + dateControl.hour + "_raw"
            val path_raw = folderPath + File.separator + fileName_raw
            val existsRaw = Utils.instance.checkFileExists(path_raw)
            if (!existsRaw) {
                val filePattern = "%m%n"
                val maxBackupSize = 250
                val maxFileSize = (2048 * 2048).toLong()
                Configure(path_raw, filePattern, maxBackupSize, maxFileSize, true)
            }

            Logger.getLogger("RAW").info(formatSimple(className, functionName, text, dateTime.getTimeFormatted(System.currentTimeMillis()), type))

        } catch (e: Exception) {
            e.printStackTrace()
        }

    }

    private fun formatSimple(SCREEN: String, FUNCTION: String, DATA: String, TIME: String, TYPE: String): String {
        return "{$SCREEN}  {$FUNCTION}  {$DATA}  {$TIME}  {$TYPE}"
    }

    private val outputPath: String
        get() = logPath + Constants.FOLDER_OUTGOING + File.separator

    private fun prepareOutputFile() {
        val outputPath = outputPath
        Log.i(TAG, "Output Path: " + outputPath)
        Utils.instance.deleteDir(File(outputPath))
        Utils.instance.createDirIfNotExists(outputPath)
    }

    private fun createZipFile(): String? {
        val zipName = "Logs.zip"

        Log.i(TAG, "Zip Name: " + zipName)

        val outputPath = outputPath
        val outputDirectory = File(outputPath)
        val filesToSend = outputDirectory.listFiles()

        if (filesToSend != null && filesToSend.isNotEmpty()) {
            Log.i(TAG, "Start Zipping Log Files.. " + filesToSend.size)
            compress.zip(filesToSend, outputPath, zipName)
            return outputPath + zipName
        }

        return null
    }

    fun getFilesForToday(folderPath: String) {

        val outputPath = outputPath
        val directory = File(folderPath)
        val files = directory.listFiles()

        if (files.isNotEmpty()) {
            for (file: File in files) {
                Utils.instance.copyFile(folderPath, file.name, outputPath)
            }
        } else {
            Log.i(TAG, "No Log Files")
        }
    }

    private val logPath: String = Environment.getExternalStorageDirectory().path + File.separator + "PLOG" + File.separator + Constants.FOLDER_LOG + File.separator


    fun getLogs(type: Int): String {

        prepareOutputFile()

        var path = ""

        when (type) {
            LOG_TODAY -> {
                Log.i(TAG, "Get All Logs for Today's Date")
                path = logPath + dateControl.today
                getFilesForToday(path)
            }

            LOG_LAST_HOUR -> {
                Log.i(TAG, "Get All Logs for Last Hour")
                path = logPath + dateControl.today
                getFilesForLastHour(path)
            }

            LOG_WEEK -> {
                Log.i(TAG, "Get All Logs for Last Week")
                getFilesForLastWeek(logPath)
            }
        }

        val output = createZipFile()
        Log.i(TAG, "Output Zip: " + output!!)
        return output
    }

    private fun getFilesForLastHour(folderPath: String): String {

        val directory = File(folderPath)
        val files = directory.listFiles()

        var lastHour = Integer.parseInt(dateControl.hour) - 1

        if (files.isNotEmpty()) {

            val found = filterFile(folderPath, files, lastHour)

            if (!found) {
                lastHour = Integer.parseInt(dateControl.hour)
                filterFile(folderPath, files, lastHour)
            }

        }
        return ""
    }

    private fun filterFile(folderPath: String, files: Array<File>, lastHour: Int): Boolean {
        var found = false

        for (i in files.indices) {
            val fileHour = extractHour(files[i].name)
            Log.i(TAG, "Last Hour: " + lastHour + " Check File Hour: " + fileHour + " " + files[i].name)
            if (fileHour == lastHour) {
                found = true
                Utils.instance.copyFile(folderPath, files[i].name, outputPath)
            }
        }

        return found
    }


    private fun extractDay(name: String): Int {
        return Integer.parseInt(name.substring(0, 2))
    }

    private fun extractHour(name: String): Int {
        return Integer.parseInt(name.substring(8, 10))
    }


    private fun getFilesForLastWeek(folderPath: String) {

        val today = Integer.parseInt(dateControl.currentDate)
        val lastWeek = Integer.parseInt(dateControl.lastWeek)

        val directory = File(folderPath)
        val files = directory.listFiles()

        if (files != null) {
            for (file in files) {
                if (file != null) {
                    if (file.isDirectory) {
                        val day = extractDay(file.name)

                        Log.i(TAG, "Today: $today  Last Week: $lastWeek File Date: $day")
                        if (day <= today && day >= lastWeek) {
                            getFilesForToday(file.path)
                        }
                    }
                }
            }
        }
    }
}
